src/iterative.F90

```diff
@@ -1371,324 +1108,170 @@ subroutine Gr_blk2csr(P,nbl,indblk,A)
+        call prealloc_mult(ESH(i+1,i), Gn(i,i+1), work3)
+        call create(factors, work1%nrow, work1%ncol)
+        factors%val = work1%val + work2%val - work3%val
 
-    !******************************************************************************
-    !Input:
-    !ESH(nbl,nbl): sparse matrices array ES-H
-    !SelfEneR(ncont): sparse matrices array with contacts Self Energy
-    !frm(ncont): Fermi diistribution value for each contact
-    !ref:  reference contact
-    !
-    !global variables needed: nbl, indblk(nbl+1), ncont, cblk(ncont), Gr(:,:)
-    !diagonal, subdiagonal, overdiagonal and in colums Gr(:,cb) where cb are
-    !layers interacting with all contacts but collector
-    !
-    !Output:
-    !Gl: sparse matrix containing G  contacts contribution
-    !*******************************************************************************
+        call prealloc_mult(gsmr(i+1), factors, Gn(i+1,i+1))
+        call destroy(factors, work1, work2, work3)
 
-    implicit none
+    end do
 
-    !In/Out
-    type(z_DNS), dimension(:,:), intent(in) :: ESH
-    type(z_DNS), dimension(:,:), intent(inout) :: Gn
-    type(z_DNS), dimension(:), intent(in) :: SelfEneR
-    type(Tstruct_info), intent(in) :: struct
-    real(dp), dimension(:), intent(in) :: frm
-    integer, intent(in) :: ref
+    call destroy_tridiag_blk(Sigma_n)
+    call deallocate_blk_dns(Sigma_n)
+ 
+    contains
+    ! Recursive calculation of Sigma_n:      
+    ! gns(i+1) = gsmr(i+1) Sigma(i+1,i+1) gsmr(i+1)^dag
+    ! Sigma(i,i) = Sigma(i,i) + Tr(i,i+1) gns(i+1) Ta(i+1,i) 
+    !                         - Tr(i,i+1) gsmr(i+1) Sigma(i+1,i) 
+    !                         - Sigma(i,i+1) gsmr^dag(i+1) Ta(i+1,i)]
+    ! 
+    subroutine calculate_sigma_n()      
+      !Work
+      type(z_DNS) :: work, gns 
+
+      ! if nbl = 1 => Sigma_n(1,1) is ready
+      if (nbl.eq.1) return
+      !g^n(nbl) = gsmr(nbl) Sigma(nbl,nbl) gsma(nbl)
+      call zdagger(gsmr(nbl),gsmrDag)
+      call prealloc_mult(gsmr(nbl), Sigma_n(nbl,nbl), work1)
+      call prealloc_mult(work1, gsmrDag, gns)
+      call destroy(gsmrDag, work1)
+
+      do i = nbl-1, 1, -1
+        !work1 = Tr(i,i+1) gns(i+1) Ta(i+1,i)
+        ! Tr(i,i+1) = ESH(i,i+1);  Ta(i+1,i) = ESH(i,i+1)^dag
+        call zdagger(ESH(i,i+1), ESHdag)
+        call prealloc_mult(ESH(i,i+1), gns, work)
+        call prealloc_mult(work, ESHdag, work1)
+        call destroy(work, gns)
+
+        !work2 = Sigma(i,i+1) gsmr^dag(i+1) Ta(i+1,i)
+        call zdagger(gsmr(i+1), gsmrDag)
+        call prealloc_mult(Sigma_n(i,i+1), gsmrDag, work)
+        call prealloc_mult(work, ESHdag, work2)
+        call destroy(work, gsmrDag, ESHdag)
+
+        !work3 = ESH(i,i+1) gsmr(i+1) Sigma(i+1,i)
+        call prealloc_mult(ESH(i,i+1), gsmr(i+1), work)
+        call prealloc_mult(work, Sigma_n(i+1,i), work3)
+        call destroy(work)
+
+        !sum of the four factors
+        Sigma_n(i,i)%val = Sigma_n(i,i)%val + work1%val - work2%val - work3%val
```

Comment:
why not accumulating factors work1, work2, work3 directly in Sigma_n one after the other as they are calculate? It'd be as readable, and can spare 2 work matrices.